lib/util.h

```diff
@@ -0,0 +1,57 @@
+#ifndef _LIB_UTIL_H_
+#define _LIB_UTIL_H_ 1
+
+// `nelems (a)` is meant to give the number of elements in `a` and should only be used when `a` is an array.
+//
+// Usage:
+//
+//   ```
+//   char buffer [512];
+//   snprintf (buffer, nelems (buffer), "The values are (%f, %f).", x, y);
+//   ```
+//
+//   This is much safer than the version using `sizeof (buffer)`
+//   (which creates a bad habit -- invitation to overflow
+//   when switching to versions of `snprintf` for `wchar_t` etc.)
+//   and is more compact than the version using `sizeof (buffer) / sizeof (buffer [0])`.
+//
+//   Also, unlike the `snprintf (buffer, 512, etc.)` version, it avoids repetition of `512`.
+//
+//   Also, unlike the following version:
+//   ```
+//   const size_t buffer_size = 512;
+//   char buffer [buffer_size];
+//   snprintf (buffer, buffer_size, "The values are (%f, %f).", x, y);
+//   ```
+//   it avoids introducing a new identifier.
+//
+// Rationale for the current implementation:
+// Using `sizeof ((a) [0])` allows `a` to be a variable with user-defined `operator[]`.
+// Using `sizeof (0 [a])` rules out user-defined `operator[]` and requires `a` to be an array or, unfortunately, a pointer.
+// Thanks to Matthew Wilson (of STLsoft fame) for this.
+//
+// TODO: We should only allow `a` to be an array (i.e. we should rule out pointers).
+
+#ifndef nelems
+    #define nelems(a) (sizeof (a) / sizeof (0 [a]))
+#endif
+
+// `azzert`, `azzert_msg`: An always-enabled (even in release builds) `assert`.
+//
+//   The purpose is to bring the crash (with its rich debugging information i.e. the core file)
+//   closer to the actual bug
+//   (by having programmers use azzertions to check pre-conditions, post-conditions, invariants etc.).
+//
+//   `azzert` is different from `assert` (in <cstdlib>) by always being enabled.
+
+[[noreturn]] void azzertionHasFailed (const char *pszFile, unsigned iLine, const char *pszFunction, const char *pszExpression, const char *pszMessage = nullptr);
+
+#ifndef azzert
+    #define azzert(b) do { if (! (b)) azzertionHasFailed (__FILE__, __LINE__, __PRETTY_FUNC__, #e); } while (0)
```

Comment:
I find "util.h" obscuring the contents of this file. This should be in either in error.h or in a file called "azzert.h".